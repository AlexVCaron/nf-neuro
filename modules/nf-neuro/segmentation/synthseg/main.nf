process SEGMENTATION_SYNTHSEG {
    tag "$meta.id"
    label 'process_single'
    label 'process_high'

    container "freesurfer/freesurfer:7.4.1"

    input:
    tuple val(meta), path(image), path(lesion) /* optional, input = [] */, path(fs_license) /* optional, input = [] */

    output:
    tuple val(meta), path("*__mask_wm.nii.gz")                , emit: wm_mask
    tuple val(meta), path("*__mask_gm.nii.gz")                , emit: gm_mask
    tuple val(meta), path("*__mask_csf.nii.gz")               , emit: csf_mask
    tuple val(meta), path("*__gm_parc.nii.gz")                , emit: gm_parc, optional: true
    tuple val(meta), path("*__resampled_image.nii.gz")        , emit: resample, optional: true
    tuple val(meta), path("*__volume.csv")                    , emit: volume, optional: true
    tuple val(meta), path("*__qc_score.csv")                  , emit: qc_score, optional: true
    path "versions.yml"                                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    def gpu = task.ext.gpu ? "" : "--cpu"
    def gm_parc = task.ext.gm_parc ? "--parc" : ""
    def robust = task.ext.robust ? "--robust" : ""
    def fast = task.ext.fast ? "--fast" : ""
    def ct = task.ext.ct ? "--ct" : ""
    def output_resample = task.ext.output_resample ? "--resample ${prefix}__resampled_image.nii.gz": ""
    def output_volume = task.ext.output_volume ?  "--vol ${prefix}__volume.csv" : ""
    def output_qc_score = task.ext.output_qc_score ?  "--qc ${prefix}__qc_score.csv" : ""
    def crop = task.ext.crop ? "--crop " + task.ext.crop: ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    cp $fs_license \$FREESURFER_HOME/license.txt

    mri_synthseg --i $image --o seg.nii.gz --threads $task.cpus $gpu $robust $fast $ct $output_resample $output_volume $output_qc_score $crop

    if [[ $task.gm_parc ]];
    then
        # Cortical grey matter parcellation
        mri_synthseg --i $image --o ${prefix}__gm_parc.nii.gz --threads $task.cpus $gpu $gm_parc $robust $fast $crop
        mri_convert -i ${prefix}__gm_parc.nii.gz --out_data_type uchar -o ${prefix}__gm_parc.nii.gz
    fi

    # WM Mask
    mri_binarize --i seg.nii.gz \
        --match 2 7 10 12 13 16 28 41 46 49 51 52 60 \
        --o ${prefix}__mask_wm.nii.gz

    # GM Mask
    mri_binarize --i seg.nii.gz \
        --match 3 8 11 17 18 26 42 47 50 52 53 54 58 \
        --o ${prefix}__mask_gm.nii.gz

    # CSF Mask
    mri_binarize --i seg.nii.gz \
        --match 4 5 14 15 24 43 44 \
        --o ${prefix}__mask_csf.nii.gz

    if [[ -f "$lesion" ]];
    then
        mri_binarize --i ${prefix}__mask_wm.nii.gz --merge $lesion --min 0.5 --o ${prefix}__mask_wm.nii.gz
    fi

    mri_convert -i ${prefix}__mask_wm.nii.gz --out_data_type uchar -o ${prefix}__mask_wm.nii.gz
    mri_convert -i ${prefix}__mask_gm.nii.gz --out_data_type uchar -o ${prefix}__mask_gm.nii.gz
    mri_convert -i ${prefix}__mask_csf.nii.gz --out_data_type uchar -o ${prefix}__mask_csf.nii.gz
    mri_convert -i ${prefix}__gm_parc.nii.gz --out_data_type uchar ${prefix}__gm_parc.nii.gz

    rm \$FREESURFER_HOME/license.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        Freesurfer: 7.4.1
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mri_synthseg -h
    mri_binarize -h
    mri_convert -h

    touch ${prefix}__mask_wm.nii.gz
    touch ${prefix}__mask_gm.nii.gz
    touch ${prefix}__mask_csf.nii.gz
    touch ${prefix}__gm_parc.nii.gz
    touch ${prefix}__resampled_image.nii.gz
    touch ${prefix}__volume.csv
    touch ${prefix}__qc_score.csv


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        Freesurfer: 7.4.1
    END_VERSIONS
    """
}
