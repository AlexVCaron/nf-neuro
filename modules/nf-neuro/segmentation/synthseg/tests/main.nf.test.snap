{
    "segmentation - synthseg - options": {
        "content": [
            "test__mask_wm.nii.gz:md5:header,07555a083d94f8f52a4227cacb0ed0b1,data,8ffff7c6810fc860e647d468f78da688",
            "test__mask_gm.nii.gz:md5:header,07555a083d94f8f52a4227cacb0ed0b1,data,33864a62b5de53801fe5a7fb708565e6",
            "test__mask_csf.nii.gz:md5:header,07555a083d94f8f52a4227cacb0ed0b1,data,f26e143ee6e50ba6de14ffa663009980",
            "test__map_wm.nii.gz:md5:header,8741ac74cc520ac7a3fc1338a8210f4d,data,1b6e41dc5ea5d9342532fc9e861c0811",
            "test__map_gm.nii.gz:md5:header,8741ac74cc520ac7a3fc1338a8210f4d,data,95e93dd0c0a5cd7ca6cf2beaf0b49394",
            "test__map_csf.nii.gz:md5:header,8741ac74cc520ac7a3fc1338a8210f4d,data,d56c1f5b9bc58681e420b7caba87d708",
            "test__aparc_aseg.nii.gz:md5:header,ff063c025479983639fd545e1e8fd690,data,adeeec0078924b2fde1cd4667a2c330d",
            "test__resampled_image.nii.gz:md5:header,728dfc231a3dfcf567ac640930f68629,data,1f02837025def8b0701f79549c2955e1",
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,b65418e9ce2ab1cce590b1f659e18c1d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T20:56:45.8845281"
    },
    "segmentation - synthseg - basic": {
        "content": [
            "test__mask_wm.nii.gz:md5:header,a0430ca63ccc3bb2b925694cb7568388,data,1b5a4d63ae2b56d9d6936313b5225b55",
            "test__mask_gm.nii.gz:md5:header,a0430ca63ccc3bb2b925694cb7568388,data,78f4ced658a08d05c2eee30f8ab87b5e",
            "test__mask_csf.nii.gz:md5:header,a0430ca63ccc3bb2b925694cb7568388,data,c9abaab9c5e047fadf928a6520929d5e",
            "test__map_wm.nii.gz:md5:header,259952abe43ff3021fb61c96b9cbebee,data,262fa0d5882b5c4c6555a5680564efcc",
            "test__map_gm.nii.gz:md5:header,259952abe43ff3021fb61c96b9cbebee,data,83999891b7edd45e9af9867b8e066321",
            "test__map_csf.nii.gz:md5:header,259952abe43ff3021fb61c96b9cbebee,data,bc4285b196045ebf3e5959da0aa65806",
            "test__seg.nii.gz:md5:header,f84c3d38470f2b2c2741e78ee95b5367,data,f68a3f27105e7fe463bb108648b11459",
            [
                "versions.yml:md5,b65418e9ce2ab1cce590b1f659e18c1d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T20:49:21.443292744"
    },
    "segmentation - synthseg - lesion": {
        "content": [
            "test__mask_wm.nii.gz:md5:header,a0430ca63ccc3bb2b925694cb7568388,data,6d99d71cff4a2f9d147c45b1ee36dce0",
            "test__mask_gm.nii.gz:md5:header,a0430ca63ccc3bb2b925694cb7568388,data,78f4ced658a08d05c2eee30f8ab87b5e",
            "test__mask_csf.nii.gz:md5:header,a0430ca63ccc3bb2b925694cb7568388,data,c9abaab9c5e047fadf928a6520929d5e",
            "test__map_wm.nii.gz:md5:header,259952abe43ff3021fb61c96b9cbebee,data,262fa0d5882b5c4c6555a5680564efcc",
            "test__map_gm.nii.gz:md5:header,259952abe43ff3021fb61c96b9cbebee,data,83999891b7edd45e9af9867b8e066321",
            "test__map_csf.nii.gz:md5:header,259952abe43ff3021fb61c96b9cbebee,data,bc4285b196045ebf3e5959da0aa65806",
            "test__seg.nii.gz:md5:header,f84c3d38470f2b2c2741e78ee95b5367,data,f68a3f27105e7fe463bb108648b11459",
            [
                "versions.yml:md5,b65418e9ce2ab1cce590b1f659e18c1d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T21:04:59.095478079"
    }
}